cpp/src/arrow/compute/kernels/vector_run_end_encode.cc

```diff
@@ -0,0 +1,672 @@
+        output_run_ends_[write_offset] = static_cast<RunEndCType>(run_end);
+        write_offset += 1;
+        // Open the new run
+        current_run_valid = valid;
+        current_run = value;
+      }
+    }
+    WriteValue(write_offset, current_run_valid, current_run);
+    DCHECK_EQ(input_length_, read_offset - input_offset_);
+    output_run_ends_[write_offset] = static_cast<RunEndCType>(input_length_);
+    return write_offset + 1;
+  }
+};
+
+template <typename RunEndType>
+Status ValidateRunEndType(int64_t input_length) {
+  using RunEndCType = typename RunEndType::c_type;
+  constexpr int64_t kRunEndMax = std::numeric_limits<RunEndCType>::max();
+  if (input_length > kRunEndMax) {
+    return Status::Invalid(
+        "Cannot run-end encode Arrays with more elements than the "
+        "run end type can hold: ",
+        kRunEndMax);
+  }
+  return Status::OK();
+}
+
+template <typename RunEndType, typename ValueType, bool has_validity_buffer>
+class RunEndEncodeImpl {
+ private:
+  KernelContext* ctx_;
+  const ArraySpan& input_array_;
+  ExecResult* output_;
+
+ public:
+  using RunEndCType = typename RunEndType::c_type;
+  using CType = typename ValueType::c_type;
+
+  RunEndEncodeImpl(KernelContext* ctx, const ArraySpan& input_array, ExecResult* out)
+      : ctx_{ctx}, input_array_{input_array}, output_{out} {}
+
+  Status Exec() {
+    const int64_t input_length = input_array_.length;
+    const int64_t input_offset = input_array_.offset;
+    const auto* input_validity = input_array_.buffers[0].data;
+    const auto* input_values = input_array_.buffers[1].data;
+
+    // First pass: count the number of runs
+    int64_t num_valid_runs = 0;
+    int64_t num_output_runs = 0;
+    if (input_length > 0) {
+      RETURN_NOT_OK(ValidateRunEndType<RunEndType>(input_length));
+
+      RunEndEncodingLoop<RunEndType, ValueType, has_validity_buffer> counting_loop(
+          input_array_.length, input_array_.offset, input_validity, input_values);
+      std::tie(num_valid_runs, num_output_runs) = counting_loop.CountNumberOfRuns();
+    }
+
+    // Allocate the output array data
+    std::shared_ptr<ArrayData> output_array_data;
+    int64_t validity_buffer_size = 0;  // in bytes
+    {
+      ARROW_ASSIGN_OR_RAISE(auto run_ends_buffer,
+                            AllocateBuffer(num_output_runs * RunEndType().bit_width(),
+                                           ctx_->memory_pool()));
+      std::shared_ptr<Buffer> validity_buffer = NULLPTR;
+      if constexpr (has_validity_buffer) {
+        validity_buffer_size = bit_util::BytesForBits(num_output_runs);
+        ARROW_ASSIGN_OR_RAISE(validity_buffer,
+                              AllocateBuffer(validity_buffer_size, ctx_->memory_pool()));
+      }
+      ARROW_ASSIGN_OR_RAISE(auto values_buffer,
+                            AllocateBuffer(bit_util::BytesForBits(
+                                               num_output_runs * ValueType().bit_width()),
```

Comment:
That's true in the case above, but `ValueType` can be `BooleanType` which fits in a single bit.

Reply:
for a `BooleanType` you should use `AllocateBitmap` instead, right?

`BytesForBits(num_output_runs * bit_width())` should be identical to `byte_width()` even in the boolean case, shouldn't it?

Reply:
`Boolean::bit_width()` returns 1:


```cpp
// BooleanType within arrow use a single bit instead of the C 8-bits layout.
int bit_width() const final { return 1; }
```

Then the `BooleanType::byte_width()` which is the default from `DataType`

```cpp
    /// \brief Returns the type's fixed byte width, if any. Returns -1
    /// for non-fixed-width types, and should only be used for
    /// subclasses of FixedWidthType
    virtual int32_t byte_width() const final {
      int32_t num_bits = this->bit_width();
      return num_bits > 0 ? num_bits / 8 : -1;
    }
  ```

evaluates to 0.

Reply:
I opened a PR adding `BytesForBits` in the filtering code after this research: https://github.com/apache/arrow/pull/34521

Reply:
oy, well that's inconvenient....

Reply:
Good news. I just added a commit that removes the need for `BytesForBits` and does the right clearing of trailing bits when allocating bitmaps.